### PR TITLE
Avoid repeatedly processing epoch transitions

### PR DESCRIPTION
**Pull request by ajsutton**
## PR Description
When creating the first block of an epoch, Teku winds up processing the epoch transition four times:
* Computing proposer duties
* `BlockFactory` creates the state at the block slot to use when getting operations to include in the block
* `BlockProposalUtil` recreates the state at the block slot as part of applying the block to get the resulting state
* `ForkChoiceUtil` reprocesses the block when it imports it

The first three of those can all be modified to reuse the same cached state.  This is done by changing our `Checkpoint` based state cache to a `SlotAndBlockRoot` based so it can cache states from arbitrary slots rather than only epoch boundaries.  Computing proposer duties looks up the state at the epoch start slot. This includes applying any empty slots if needed.
Both `BlockFactory` and `BlockProposalUtil` are invoked as a result of the same validator client call to create the unsigned block - the state from the new block's slot can be retrieved from cache then passed in to both of them to reuse.

Note that the caching works out correctly even when we request a state at the first slot of an epoch prior to importing the block because that's cached under `(start_slot, previous_block_root)` and then if a block is later imported for that slot it's cached under `(start_slot, new_block_root)` so we can't pollute the cache by looking a state for a slot too early.

## Documentation

- [x] I thought about documentation and added the `documentation` label to this PR if updates are required.

### CODE REVIEW

File: ethereum/core/src/test/java/tech/pegasys/teku/core/stategenerator/StateAtSlotTaskTest.java [hunk 1]
```diff
@@ -26,19 +26,21 @@
...
 import tech.pegasys.teku.datastructures.state.Checkpoint;
 import tech.pegasys.teku.datastructures.util.BeaconStateUtil;
 import tech.pegasys.teku.infrastructure.async.SafeFuture;
 import tech.pegasys.teku.infrastructure.unsigned.UInt64;
 
-class CheckpointStateTaskTest {
-  private static final UInt64 EPOCH = UInt64.valueOf(2);
-  private static final UInt64 EPOCH_START_SLOT = BeaconStateUtil.compute_start_slot_at_epoch(EPOCH);
+class StateAtSlotTaskTest {
+  private static final UInt64 EPOCHx = UInt64.valueOf(2);
```
Comment (rolfyone) [inline ethereum/core/src/test/java/tech/pegasys/teku/core/stategenerator/StateAtSlotTaskTest.java]: typo? (EPOCH + x)

File: validator/coordinator/src/main/java/tech/pegasys/teku/validator/coordinator/BlockFactory.java [hunk 2]
```diff
@@ -87,7 +94,12 @@ public BeaconBlock createUnsignedBlock(
     }
 
     // Collect attestations to include
-    final BeaconState blockSlotState = stateTransition.process_slots(blockPreState, newSlot);
+    final BeaconState blockSlotState;
+    if (maybeBlockSlotState.isPresent()) {
+      blockSlotState = maybeBlockSlotState.get();
+    } else {
+      blockSlotState = stateTransition.process_slots(blockPreState, newSlot);
+    }
```
Comment (rolfyone) [inline validator/coordinator/src/main/java/tech/pegasys/teku/validator/coordinator/BlockFactory.java:102]: might be able to use `orElseGet` instead here?

Comment (ajsutton) [reply validator/coordinator/src/main/java/tech/pegasys/teku/validator/coordinator/BlockFactory.java:102]: ah I remember trying that before now.... `process_slots` can throw checked exceptions so it doesn't work well with `orElseGet`.

File: ethereum/core/src/test/java/tech/pegasys/teku/core/stategenerator/StateAtSlotTaskTest.java (hunk 1 above)
Comment (ajsutton) [reply ethereum/core/src/test/java/tech/pegasys/teku/core/stategenerator/StateAtSlotTaskTest.java]: Fixed.

**Review by rolfyone — APPROVED** [review]
LGTM

**Pull request merged**